microsoft/testsuites/network/netinterface.py

```diff
@@ -0,0 +1,120 @@
+from lisa.operating_system import CoreOs
+from lisa.tools import Dhclient, Uname, Wget
+from lisa.util import constants
+
+
+@TestSuiteMetadata(
+    area="network",
+    category="functional",
+    description="""
+    This test suite validates basic functionalities of Network Interfaces.
+    """,
+)
+class NetInterface(TestSuite):
+    RELOAD_NETVSC_TEST_COUNT = 10
+    NET_IFUP_IFDOWN_TEST_COUNT = 4
+    DHCLIENT_WAIT_TIME = 15
+
+    @TestCaseMetadata(
+        description="""
+            This test case verifies if synthetic network interface can be
+            brought up and brought down gracefully.
+
+            Steps:
+            1. Validate netvsc isn't built-in already. If it is then skip the test.
+            2. Unload and load netvsc module multiple times in a loop.
+            3. Change nic state to up and down multiple times.
+                Each time after "up" state, verify ip address is assigned to nic
+                and internet is accessible via nic.
+
+        """,
+        priority=1,
+        requirement=simple_requirement(
+            min_nic_count=1,
+            network_interface=schema.NetworkInterfaceOptionSettings(
+                data_path=schema.NetworkDataPath.Synthetic,
+            ),
+        ),
+    )
+    def validate_network_interface_ifup_ifdown(self, node: Node, log: Logger) -> None:
+        self._validate_netvsc_built_in(node)
+        network_interface_feature = node.features[NetworkInterface]
+
+        # Test loading and unloading netvsc driver
+        test_count = 0
+        while test_count < self.RELOAD_NETVSC_TEST_COUNT:
+            test_count += 1
+            # Unload and load hv_netvsc
+            network_interface_feature.reload_netvsc()
+
+        # Clean all dhclient processes, get IP & try ping
+        node_nic_info = Nics(node)
+        node_nic_info.initialize()
+        default_nic = node_nic_info.default_nic
+        default_route = node_nic_info.default_nic_route
+
+        assert_that(default_nic).is_not_none()
+        assert_that(default_route).is_not_none()
+
+        test_count = 0
+        while test_count < self.NET_IFUP_IFDOWN_TEST_COUNT:
+            log.info(f"NET IFUP IFDOWN Test Iteration : {test_count}")
+            node_nic_info.reset_nic_state(default_nic)
+            if not node_nic_info.default_nic:
+                # Add default route if missing after running ip link down/up
+                node.execute(f"ip route add {default_route}", shell=True, sudo=True)
+            if not node_nic_info.nics[default_nic].ip_addr:
+                node.execute("kill $(pidof dhclient)", shell=True, sudo=True)
+                dhclient = node.tools[Dhclient]
+                dhclient.release()
+                dhclient.renew()
+
+            wait_start = time.time()
+            wait_delta = 0.0
+            while wait_delta < self.DHCLIENT_WAIT_TIME:
+                time.sleep(2)
+                if node_nic_info.nics[default_nic].ip_addr:
```

Comment:
Is there any logic to trigger to retrieve the ip_addr? I'm not sure if it will be changed in each try.

Reply:
Hmm let me confirm though I believe it shouldn't change

Reply:
we need to call some commands to refresh the ip address.

Reply:
> we need to call some commands to refresh the ip address.

you mean refresh the ip address in the nic object? reset_nic_state() does that by calling _get_interface_info() again after setting nic state down and up again.

Reply:
I mean, in the `while` loop, it's checking if the ip_addr is assigned. If it's assigned, it will return. But in the loop, there is no command to retrieve ip address from the system. So the ip_addr is not changed for each time it rans. The interval loop doesn't work actually. if the ip_addr is not set in first time, it will wait the whole time. It won't return earlier.